[stdlib] Speed up String(largeCharacter)

dabrahams opened the pull request:
There's no need to do grapheme breaking again here.  Not sure why it was needed in the first place, but it was ostensibly to fix <rdar://19803733> which appears to still work fine with this change.

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
OK, so our tests are at least dependent on the old code; let's see if this improves the benchmarks.  Maybe we can find a way to avoid the overhead.

dabrahams commented:
@milseman @airspeedswift This demonstrates that if we can remember when a `Character` contains a single grapheme cluster we can get a 3x speedup on SuperChars.

milseman commented on stdlib/public/core/Character.swift:
```diff
@@ -150,7 +150,7 @@ public struct Character :
   ///     let oBreve: Character = "o\u{306}"
   ///     print(oBreve)
-  ///     // Prints "ŏ"
+  ///     // Prints "ŏ"
```
Err, what is this change?

dabrahams replied:
Search me.  Did something re-encode the text?

milseman commented on stdlib/public/core/Character.swift:
```diff
@@ -398,8 +398,7 @@ extension String {
     case let .large(value):
-      let buf = String(_StringCore(_StringBuffer(value)))
-      self = buf[buf.startIndex..<buf.index(after: buf.startIndex)]
+      self = String(_StringCore(_StringBuffer(value)))
```
This change contradicts the comments about what `value` is, at least.

```
  internal enum Representation {
    // A _StringBuffer whose first grapheme cluster is self.
    // NOTE: may be more than 1 Character long.
    case large(_StringBuffer._Storage)
    case small(Builtin.Int63)
  }
```

milseman replied:
(I have no attachment to any particular set of assumptions or tradeoffs, just that they be consistent)

dabrahams replied:
...which explains why the tests failed.  Clearly we can't accept the PR as-is, but it was not meant to be a candidate for merging; just a demonstration of possible speedup.  The question is, what are we gaining by putting off finding the end of the first grapheme?  I don't know.  If there's time for me to land my Character rewrite, we'll find out which is faster eventually anyhow, because it does not delay like this.

milseman reviewed (dismissed):
If value cannot be more than 1 grapheme, then please correct the comment and check for any scattered assumptions about that.

Commit:
[stdlib] A Character's buffer should only contain 1 grapheme

This speeds up construction of a String from large Character representations,
and various other operations that would otherwise require additional grapheme
breaking just to interpret the Character.

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please benchmark

milseman commented on stdlib/public/core/Character.swift:
```diff
@@ -137,15 +135,16 @@ public struct Character :
   /// Creates a character with the specified value.
   ///
-  /// Do not call this initializer directly. It is used by the compiler when
+  /// Do not call this initalizer directly. It is used by the compiler when
```
The mysterious case of the missing i

dabrahams replied:
Patches welcome! ;-)

milseman commented on stdlib/public/core/Character.swift:
```diff
@@ -399,8 +399,7 @@ extension StringProtocol {
     case let .large(value):
-      let buf = String(_StringCore(_StringBuffer(value)))
-      self = Self(buf[buf.startIndex..<buf.index(after: buf.startIndex)])
+      self = Self(String(_StringCore(_StringBuffer(value))))
```
Do you think StringProtocol should have an (internal) init directly from a _StringCore?

dabrahams replied:
How would you implement that?

dabrahams replied:
That's probably more appropriate for `_SwiftStringView`.

milseman commented on stdlib/public/core/Character.swift:
```diff
@@ -203,7 +202,8 @@ public struct Character :
-       native.start == s._core._baseAddress! {
+      native.start == s._core._baseAddress!,
+      native.usedCount == s._core.count {
       _representation = .large(native._storage)
```
So in theory, until we reform Character's small representation to be inline UTF16, will this pessimize code that iteratively peeks at the first character of a String/Substring? That is, are we speeding up String construction from a single Character at the cost of non-ASCII consumer loops?

dabrahams replied:
I doubt it.  Doing just about anything with that character, including comparing it to another character, requires grapheme breaking again if we don't do this.

milseman reviewed (commented):
Looks good in combination with a revamp of Character's small form. In isolation, I'm concerned it might pessimize some patterns.

milseman commented:
@shahmishal does the benchmark builder need to be cleaned?
```
FAILED: lib/swift/macosx/x86_64/libswiftCloudKit.dylib 
...
ld: library not found for -lswiftContacts for architecture x86_64
```

milseman commented:
@swift-ci please clean test

dabrahams commented:
on Mon May 15 2017, Michael Ilseman <notifications-AT-github.com> wrote:

> milseman commented on this pull request.
>
> Looks good in combination with a revamp of Character's small form. In
> isolation, I'm concerned it might pessimize some patterns.

What patterns, and why are you concerned?

-- 
-Dave

milseman commented on stdlib/public/core/Character.swift:
```diff
@@ -203,13 +202,12 @@ public struct Character :
-    var nativeString = ""
-    nativeString.append(s)
-    _representation = .large(nativeString._core.nativeBuffer!._storage)
+    _representation = .large(_StringCore(s.utf16).nativeBuffer!._storage)
```
Then don't you have to do grapheme breaking here? That is, why would you guard against the same `count` above but not here?

dabrahams replied:
It's already guaranteed to be one character, based on usage.  But a _sanityCheck would be good.

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please benchmark

milseman commented on validation-test/compiler_crashers_fixed/28670-reftype-hastypeparameter-cannot-have-a-dependent-type-here.swift:
```diff
@@ -6,5 +6,15 @@
+
+// REQUIRES: deterministic-compiler
+// see https://bugs.swift.org/browse/SR-4894
+>>>>>>> 38d062e9132... [stdlib] A Character's buffer should only contain 1 grapheme
```
Looks like a merge conflict

dabrahams replied:
@milseman Gah!  Thought I took care of that.

dabrahams commented:
@swift-ci Please test os x platform

dabrahams commented:
@swift-ci Please test os x platform

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

milseman commented:
@swift-ci please test OS X Platform

milseman commented:
@swift-ci please test OS X Platform

Commit:
Merge branch 'master' into string-init-from-character-speedup

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please smoke test

dabrahams commented:
Taking for the 3x speedup in SuperChars.  Will examine the 25% slowdown in StrToInt separately.
